### DISCUSSION TITLE
Welcome to afip.ts Discussions!

### DISCUSSION THREAD

**Discussion by ralcorta** [valiulab/afip.ts#4] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by juanmacugat** [comment]
hola @ralcorta! en primer lugar, muy buena la contribucion! me parece genial hasta el momento.
te queria consultar, quizas estas al tanto,  si las facturas emitidas se pueden descargar en formato pdf mediante algun servicio de afip? desconozco si afip normalizo la generacion de los pdf mediante algun ws, o bien, con los datos del comprobante, cada empresa debe generar el pdf con un template de factura propio.
Lo que tengo entendido es que esta regulado y cada factura debe incluir un codigo de barras y codigo QR, pero no he visto ningun ws que ofrezca esto.
De hecho, los comprobantes en linea desde AFIP permiten ver y descargar un documento PDF, pero no se como hacer para aquellos generados por WS.
Muchas gracias de antemano!

**Reply by ralcorta** [reply]
@juanmacugat Muchas gracias! Estuve un poco apretado de trabajo y no pude ver si habia noticias por aca!
Actualmente no estoy enterado de ningun WS que te permita descargar un archivo PDF de la factura emitida. Tengo entendido que por este motivo es que hay numerosos packages destinados a la creacion de PDFs a base de los datos de la factura.
De ser algo que se este buscando podriamos ver añadir esa Feature al proyecto.

**Reply by ralcorta** [reply]
@juanmacugat Me encuentro actualmente trabajando en un branch sobre este mismo tema, el de crear un PDF para las facturas. Va a tardar un poco en salir ya que me encuentro un poco justo de tiempos pero voy a ir metiendo mano en cuanto pueda.

Branch: [feature/pdf](https://github.com/ralcorta/afip.ts/tree/feature/pdf)

**Comment by ataschz** [comment]
Hola gente, muy buen proyecto! Estaba investigando librerias que ayuden a interfacear con AFIP y encontra primero AfipJS y luego esta. Entiendo que esta es fork de la otra, pero que ahora esta siguiendo su propio camino, no es asi?

Por otro lado, ahora en AFIP.JS, agregaron un SaaS, lo cual, entiendo que es algo "natural" pero no termino de entender si se puede o no utilizar sin el dichoso `access_token` que ellos te proveen (y cobran).
Queria preguntar que planes tenes para esta lib @ralcorta, si es que a futuro pensas hacer lo mismo (poner un SaaS y paywall) o si tu idea es que esta libreria sea lo mas abierta posible?

**Reply by ralcorta** [reply]
Buenos dias @ataschz !

Este proyecto nació con la idea de ser una opción tipada y de código abierto, y deseo que siempre se mantenga así. Tengo planes de brindar un servicio SaaS para aquellos que no quieran o no sepan implementar esta solución. Este servicio será la misma solución, pero con una plataforma detrás que ofrecerá soporte adicional a los desarrolladores.

Actualmente, tengo una rama en la que estoy trabajando en una nueva función para generar PDFs de las facturas. Además, estoy migrando el proyecto a un monorepo, lo que permitirá agregar más librerías relacionadas con los servicios de AFIP, facilitando así el trabajo con ellos.

**Como resumen:**
Nunca va a haber un paywall por el código de este proyecto ni por los paquetes de código abierto que agregue a su alrededor.
Siempre se va a mantener open source.
Se van a crear librerías adicionales para ampliar la experiencia de desarrollo.
Se va a crear un servicio basado en el mismo código para aquellos que no sepan, no quieran o utilicen otra tecnología que no sea TypeScript.

**Reply by ataschz** [reply]
Excelente @ralcorta! Voy a probar le paquete y si tengo feedback o puedo aportar, seguramente me veas por acá.

**Reply by ataschz** [reply]
A proposito. Yo tengo mucha experiencia trabajando con monorepos, si tenes una duda o puedo darte una mano en eso, avisa!

**Comment by bertonefederico1** [comment]
Buenas noches, en primer lugar agradezco la contribución de este proyecto, es genial!
Por otro lado consulto, hay algun servicio generic al cual pueda pasarle el nombre del metodo? ya que continuamente van sacando nuevos métodos que no estan implementados en esta lib. Desde ya muchas gracias

**Reply by ralcorta** [reply]
Buenos dias @bertonefederico1 !
Acabo de realizar un cambio en la arquitectura del paquete y en la forma en que vamos a trabajar el repositorio. Ahora pasó a ser un monorepo, donde vamos a ir incorporando distintas features, como la generación de PDFs y otras funcionalidades.

Ya dejé un servicio genérico para que puedas probar. Usalo tranquilo y contame si encontrás algo para mejorar.
¡Gracias por el aporte!

**Reply by bertonefederico1** [reply]
Buenas tardes, me econtré con este problema, al usar el servicio generico del ejemplo, me arroja este error

Error: WSDL ws_sr_padron_a4 not found

este es el codigo
return await arca.genericService.call("ws_sr_padron_a4", "dummy", {});

**Reply by ralcorta** [reply]
Hola @bertonefederico1 

sí eso ya está resuelto a partir de `@arcasdk/core@2.0.0` (la última en npm es `2.0.2`).

En 1.x `genericService.call()` usaba el alias WSAA como nombre de WSDL (`ws_sr_padron_a4`), pero el WSDL embebido se llama `ws_sr_padron_a4.wsdl`, de ahí el `Error: WSDL ws_sr_padron_a4 not found`

En 2.x el mismo llamado funciona:

```ts
import { Arca, ArcaServiceNames } from "@arcasdk/core";

const result = await arca.genericService.call(
  ArcaServiceNames.WSSR_PADRON_FOUR, // "ws_sr_padron_a4"
  "dummy",
  {},
);
```

Si lo que necesitás es padrón A4, conviene el servicio tipado:

```ts
const status = await arca.registerScopeFourService.getServerStatus();
const taxpayer = await arca.registerScopeFourService.getTaxpayerDetails(20111111111);
```

Actualizá con:

```bash
npm i @arcasdk/core@latest
```

Docs: [servicio genérico](https://ralcorta.github.io/arcasdk/services/generic-service.html) y [padrón alcance 4](https://ralcorta.github.io/arcasdk/services/consulta_padron_alcance_4.html).

**Comment by Pablushka** [comment]
Hola a todos, gracias por el proyecto y la filosofía. Yo no estoy usando  Arca SDK, lo descubrí demasiado tarde para mi desgracia. Sin embargo creo que puedo colaborar con algunas cosas en el futuro. Y digo que llegué tarde porque ya empecé  una implementación parecida pero en lenguaje Rust. Mi proyecto incluye también un server con caché  LRU para almacenar los datos de conexión, tokens, etc. Estoy contento porque anda super rápido y es muy estable, incluso a pesar de Arca. Todavía no lo liberé porque me faltan bocha de servicios por implementar...

Este proyecto (Arca SDK) me ayudó a entender muchas cosas que había arrancado mal. El mayor problema que tengo es que no se cómo hacer pruebas con empresas delegadas, qué cuits de prueba usar, me da la sensación que los servidores de homologación de arca están desactulizados y sólo sirven para pruebas básicas... quizás me podrían ayudar a enteder como hacer estas pruebas. 

Por otro lado, si es de su interés podría colaborar para ayudarlos a hacer esta lib en Rust para que tengan mayor alcance, otra opción. 

Saludos y gracias nuevamente por compartir!

**Comment by enegraso** [comment]
<img width="454" height="660" alt="image" src="https://github.com/user-attachments/assets/f629fa17-341a-4b6b-9808-23ea970013b7" />
Hola gente! este proyecto está genial, lo implementé en mi backend, por ahora solo para mi; y, hasta que haga mi facturdor web, lo consumo desde mi aplicacion hecha en c#, donde armo el qr del CAE y genero el pdf con con report viewer.
Solo tendre que convertir la palabra arca a imagen para que no me pierda la fuente.
Muchas gracias por este aporte.

**Reply by ralcorta** [reply]
Muchas gracias por tu aporte y me alegro que lo estes pudiendo usar en tus proyectos! 
Hay un nuevo package dentro del mono repo que te permite hacer los pdf de las facturas: https://github.com/ralcorta/arcasdk/tree/main/packages/pdf
